File: Language.Implementation.Patterns/cxx/ch02_basic_parsing_patterns/p02_lexer/run/LexerRun.cpp
#include <string>
#include <iostream>
#include <fstream>
#include <sstream>

#include "Token.hpp"
#include "ListLexer.hpp"

int main(int argc, char* argv[]) {
    std::string file_path = "test.txt";

    if (argc == 2) {
        file_path = argv[1];
    }

    std::ifstream ifile{file_path};
    if (!ifile) {
        std::cerr << "open file \"" << file_path << "\" failed\n";
        return 1;
    }
    std::stringstream buf;
    buf << ifile.rdbuf();

    ListLexer lexer{buf.str()};
    auto t = lexer.nextToken();
    while (t.type != Token::EOF_TYPE) {
        std::cout << "=> " << t << "\n";
        t = lexer.nextToken();
    }
}

File: Language.Implementation.Patterns/python/ch02_basic_parsing_patterns/p02_lexer/LexerRun.py
import sys

from Token import Token
from Lexer import Lexer
from ListLexer import ListLexer

def main():
    file_path = "test.txt"
    if len(sys.argv) > 1:
        file_path = sys.argv[1]

    with open(file_path) as f:
        lexer = ListLexer(f.read())
        t = lexer.nextToken()
        while t.type != ListLexer.T_TYPE_EOF:
            print("=> {}".format(t))
            t = lexer.nextToken()

main()
